### PR TITLE
Import

### PR DESCRIPTION
**Pull request by babelouest**
Pull request retry
I did a git rebase before sending the pull request, hope it looks better now.

### CODE REVIEW

**Comment by tanghus** [comment]
> I did a git rebase before sending the pull request, hope it looks better now.

Sorry, but no it doesn't. Look at the number of changed files. I guess you haven't changed 313 files? ;) 

I'll have a look at it tonight (CET)

**Comment by babelouest** [comment]
@tanghus , no, I haven't changed all those files, although most of the reported changes are ACL changes. I don't know why I did that, but I changed most of the access control from 0644 to 0755...
Can I make git to skip it when it's only ACL changes ?

**Comment by tanghus** [comment]
> Can I make git to skip it when it's only ACL changes ?

Yes, but not github. I suggest you make a new clone of the repo, make a new branch, and take the files you've changed and manually copy them from the old clone to the new.

But wait till tomorrow; I'll see if I can sort it out.

**Comment by tanghus** [comment]
@babelouest Looking at your commits I figured it was the one with the `WTF ?` comment 257d082  ;)

**Comment by tanghus** [comment]
@babelouest I think I've resolved it. Please check in the Files changed tab and see if it looks correct.

**Comment by babelouest** [comment]
@tanghus , looks correct to me, thanks !

**Comment by tanghus** [comment]
@babelouest Sorry I haven't been able to review this yet. I'm a bit tight on time in the beginning of the week though. Just to let you know I haven't forgotten :)

File: formats/import_vcard_yahoo_connector.xml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		</vcard_entry>
+	</import_entry>
+	
+	<import_entry property="X-IM" enabled="true">
+		<vcard_entry property="IMPP" >
+		</vcard_entry>
+	</import_entry>
+	
+	<import_entry property="X-PRIMARY-PHONE" enabled="true">
+		<vcard_entry property="PHONE" type="HOME">
```
Comment (tanghus) [inline formats/import_vcard_yahoo_connector.xml:96]: @babelouest Can you add more than one type here? `PRIMARY` indicates it should also have type `PREF`. Or am I wrong here?

File: lib/connector/importcsvconnector.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		$ignore_first_line = (isset($this->configContent->import_core->ignore_first_line)
+								&& (((string)$this->configContent->import_core->ignore_first_line) == 'true')
+									|| ((string)$this->configContent->import_core->ignore_first_line) == '1');
+
+		$titles = false;
+
+		$lines = array();
+
+		$index = 0;
+		foreach($csv as $line)
```
Comment (tanghus) [inline lib/connector/importcsvconnector.php]: `foreach($csv as $line) {`

File: lib/connector/importcsvconnector.php [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	 *         1 - 0.5*(number of untranslated elements/total number of elements)
+	 * The more the first element has untranslated elements, the more the result is close to 0.5
+	 */
+	public function getFormatMatch($file) {
+		// Examining the first element only
+		$partsAndTitle = $this->getSourceElementsFromFile($file, 1);
+		$parts = $partsAndTitle[0];
+		$titles = $partsAndTitle[1];
+
+		if (!$parts || ($parts && isset($this->configContent->import_core->expected_columns) && count($parts[0]) != (string)$this->configContent->import_core->expected_columns)) {
```
Comment (tanghus) [inline lib/connector/importcsvconnector.php]: For readbility:

``` php
if (!$parts || ($parts && isset($this->configContent->import_core->expected_columns) 
    && count($parts[0]) != (string)$this->configContent->import_core->expected_columns)
) {
```

File: lib/connector/importvcardconnector.php [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	 * @param $input the input file to import
+	 * @param $limit the number of elements to return (-1 = no limit)
+	 * @return array of strings
+	 */
+	public function getElementsFromInput($file, $limit=-1) {
+
+		$parts = $this->getSourceElementsFromFile($file, $limit);
+		
+		$elements = array();
+		foreach($parts as $part)
```
Comment (tanghus) [inline lib/connector/importvcardconnector.php]: `foreach($parts as $part) {`

File: lib/importmanager.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+use OCA\Contacts\Connector\ImportVCardConnector;
+use OCA\Contacts\Connector\ImportLdifConnector;
+use OCA\Contacts\Addressbook;
+
+/**
+ * Manages the import with basic functionalities
+ */
+class ImportManager {
+	
+	public function __construct() {
```
Comment (tanghus) [inline lib/importmanager.php]: If not used just remove it.

**Comment by tanghus** [comment]
Just a few coding style issues. Full test later :)

File: formats/import_vcard_yahoo_connector.xml (hunk 1 above)
Comment (babelouest) [inline formats/import_vcard_yahoo_connector.xml:96]: I'm not sure I understand your question.
Here, `X-PRIMARY-PHONE` stands for the property name that we're looking for, to be replace by the one below (`PHONE type:HOME`).
If you want to convert other properties, just add an `<import_entry>` tag with the `<vcard_entry>` tag for the result.
What do you mean by `PREF` ? is it a VCard property ?

**Comment by babelouest** [comment]
no problem :)

File: formats/import_vcard_yahoo_connector.xml (hunk 1 above)
Comment (tanghus) [inline formats/import_vcard_yahoo_connector.xml:96]: PREF is a standard type for the preferred property if there are more than one. Used for e.g TEL, EMAIL, URL, ADR etc. Like in `TEL;TYPE=HOME,PREF:+1-919-676-9564` or `TEL;TYPE=HOME;TYPE=PREF:+1-919-676-9564`. Not sure if Yahoo has more than one PHONE type though?

Comment (babelouest) [inline formats/import_vcard_yahoo_connector.xml:96]: so if it's a type, you can use this:
`<vcard_entry property="PHONE" type="HOME,PREF">`

Comment (tanghus) [inline formats/import_vcard_yahoo_connector.xml:96]: Ah OK. Excellent!

**Comment by tanghus** [comment]
@babelouest First test with a Google CSV:
1. `csv_gmail/import/upload` returned `count: 0` even though there was 62 contacts and they were apparently recognized.
2. None of the contacts were parsed correctly, but all resulted in almost similar entries. See below.
3. All entries got a BDAY of current date.
4. CATEGORIES was recognized, but parsed incorrectly: `* My Contacts ::: * Starred`

Not in "My contacts":

```
BEGIN:VCARD
VERSION:3.0
X-UNKNOWN-ELEMENT;TYPE=ÿþName:
N:;;;;
X-UNKNOWN-ELEMENT;TYPE=Yomi Name:
X-PHONETIC-FIRST-NAME:
X-UNKNOWN-ELEMENT;TYPE=Additional Name Yomi:
X-PHONETIC-LAST-NAME:
NICKNAME:
X-UNKNOWN-ELEMENT;TYPE=Short Name:
X-UNKNOWN-ELEMENT;TYPE=Maiden Name:
BDAY:2014-03-04
GENDER:
X-UNKNOWN-ELEMENT;TYPE=Location:
X-UNKNOWN-ELEMENT;TYPE=Billing Information:
X-UNKNOWN-ELEMENT;TYPE=Directory Server:
X-UNKNOWN-ELEMENT;TYPE=Mileage:
X-UNKNOWN-ELEMENT;TYPE=Occupation:
X-UNKNOWN-ELEMENT;TYPE=Hobby:
X-UNKNOWN-ELEMENT;TYPE=Sensitivity:
X-UNKNOWN-ELEMENT;TYPE=Priority:
X-UNKNOWN-ELEMENT;TYPE=Subject:
NOTE:
CATEGORIES:
EMAIL1.EMAIL;TYPE=* :
FN:
UID:20140304T175950.59641b4556@localhost
REV:2014-03-04T18:00:09+00:00
PRODID:-//ownCloud//NONSGML Contacts 0.3.0.1//EN
END:VCARD
```

In "My contacts":

```
BEGIN:VCARD
VERSION:3.0
X-UNKNOWN-ELEMENT;TYPE=ÿþName:
N:;;;;
X-UNKNOWN-ELEMENT;TYPE=Yomi Name:
X-PHONETIC-FIRST-NAME:
X-UNKNOWN-ELEMENT;TYPE=Additional Name Yomi:
X-PHONETIC-LAST-NAME:
NICKNAME:
X-UNKNOWN-ELEMENT;TYPE=Short Name:
X-UNKNOWN-ELEMENT;TYPE=Maiden Name:
BDAY:2014-03-04
GENDER:
X-UNKNOWN-ELEMENT;TYPE=Location:
X-UNKNOWN-ELEMENT;TYPE=Billing Information:
X-UNKNOWN-ELEMENT;TYPE=Directory Server:
X-UNKNOWN-ELEMENT;TYPE=Mileage:
X-UNKNOWN-ELEMENT;TYPE=Occupation:
X-UNKNOWN-ELEMENT;TYPE=Hobby:
X-UNKNOWN-ELEMENT;TYPE=Sensitivity:
X-UNKNOWN-ELEMENT;TYPE=Priority:
X-UNKNOWN-ELEMENT;TYPE=Subject:
NOTE:
CATEGORIES:* My Contacts ::: * Starred
EMAIL1.EMAIL;TYPE=* Home:
FN:
UID:20140304T175950.84593cab0d@localhost
REV:2014-03-04T17:59:55+00:00
PRODID:-//ownCloud//NONSGML Contacts 0.3.0.1//EN
END:VCARD
```

Something to start with ;)

**Comment by babelouest** [comment]
OK, that's a problem indeed :p
Can you send me the csv file you used ?

**Comment by tanghus** [comment]
> Can you send me the csv file you used ?

Done

**Comment by tanghus** [comment]
BTW when the contact is in the `Starred` group it should be added to Favorites instead.

**Comment by babelouest** [comment]
OK, you tried to trick me with a UTF16 file where I expect a UTF8 file ! Damn you ! ;)
This has been fixed in the code.

About the `CATEGORIES` issue, I see two problems there.
The fact that when there are several categories they are separated with `:::`and also that every category is prefixed with `*`, no problem, I can add a separator and remove the prefix.
But fixing `when the contact is in the Starred group it should be added to Favorites instead` needs more info.
Is "Favourites" a `CATEGORIES` value ?

File: lib/connector/importcsvconnector.php (hunk 2 above)
Comment (babelouest) [inline lib/connector/importcsvconnector.php]: ok

File: lib/connector/importcsvconnector.php (hunk 3 above)
Comment (babelouest) [inline lib/connector/importcsvconnector.php]: ok

File: lib/connector/importvcardconnector.php (hunk 4 above)
Comment (babelouest) [inline lib/connector/importvcardconnector.php]: ok

File: lib/importmanager.php (hunk 5 above)
Comment (babelouest) [inline lib/importmanager.php]: ok

**Comment by tanghus** [comment]
> OK, you tried to trick me with a UTF16 file where I expect a UTF8 file ! Damn you ! ;)

I wasn't even aware of that. Just exported it from google contacts :)

> Is "Favourites" a CATEGORIES value ?

It is kinda like the Starred group in google contacts:

For CATEGORIES you have to - besides converting them to vCard CATEGORIES - index them using the `OCP\Tags` lib. This is done automatically when just adding them, but you should treat `Starred` as Favorites in the contacts app.
In the controller that is very easy:

``` php
$tagMgr = $this->server->getTagManager()->load('contact');
$tagMgr->addToFavorites($id);
```

But you may wanna pass [$this->server](https://github.com/owncloud/core/blob/master/lib/private/server.php) as an argument to the import class instead..?
Mentioning @jancborchardt just to be sure that Gmail `Starred` contacts should be filtered out and converted to favorites. And what to do with the `My Contacts` group? I don't quite get what google thinks is the difference between the two?

**Comment by tanghus** [comment]
> OK, you tried to trick me with a UTF16 file where I expect a UTF8 file ! Damn you ! ;)
> This has been fixed in the code.

How do you do that? I'm not sure I have come across any; at least I haven't noticed :P

**Comment by babelouest** [comment]
```
I wasn't even aware of that. Just exported it from google contacts :)
```

No problem, to be sure, I translate the file into UTF8 all the time now

```
It is kinda like the Starred group in google contacts:
```

Yeah but I mean, does the csv value `* My Contacts ::: * Starred` need to be transformed into the VCard property `CATEGORIES:My Contacts,Favourites` ?

**Comment by tanghus** [comment]
> Yeah but I mean, does the csv value \* My Contacts ::: \* Starred need to be transformed into the VCard property CATEGORIES:My Contacts,Favourites ?

I would say it should be transformed into `CATEGORIES:My Contacts` plus adding the contact id to favorites using the call I quoted above.
The OCP\Tags lib (formerly OC_VCategories) was originally made for fast access access to CATEGORIES in VCARD, VEVENT, VJOURNAL etc. but I have since removed the VObject specific code from it, to make it a more general "Tagging" lib. See e.g. https://github.com/owncloud/core/issues/3812
The groups in contacts app utilizes it and calendar app uses it for basic CATEGORIES editing.
For use with VObjects it requires an amount of data duplication to keep the content of the objects CATEGORIES property in sync with the index, but the code required for doing that is well made up for by the speed gained.

**Comment by babelouest** [comment]
I will put this in the xml format and change the code as needed to make the changes:

``` xml
  <import_entry name="Group Membership" enabled="true" separator=":::" remove="*">
    <vcard_favourites value="Starred">
    </vcard_favourites>
    <vcard_entry property="CATEGORIES">
    </vcard_entry>
  </import_entry>
```

Then the parsing will be modified like this:

``` php
$csv_value = str_replace($remove, '', $csv_value);
$values = explode($separator, $csv_value);
foreach($values as $value) {
    if ($value == $vcard_favourites) {
        $tagMgr->addToFavorites($id);
    } else {
        // Parse normally
    }
}
```

**Comment by tanghus** [comment]
Nice!

**Comment by babelouest** [comment]
Or even better:

``` php
    if ($value == $vcard_favourites) {
        $property = \Sabre\VObject\Property::create("X-FAVOURITES", "yes");
        $vcard->add($property);
    } else {
        // Parse normally
    }
```

Then, the `$tagMgr->addToFavorites($id);` will be done in the `importController` class, right after the `$addressbook->add($vcard)` if the VCard has the `X-FAVOURITES` property set to `yes`.

**Comment by jancborchardt** [comment]
There still are a lot of merge commits and 2 "WTF?" commits. ;) Would be cool to clean the history up a bit, otherwise no one from outside knows what happened to the code.

**Comment by babelouest** [comment]
@jancborchardt , I agree, but how can I clean the history ? Can I make a PR with just the last commit and not the previous ones for example ? (I have googled it and couldn't find an answer...)

**Comment by jancborchardt** [comment]
I guess easiest here would be to make a fresh branch off master and do the changes again manually. I know it's a pain, but in this case (looking at the diff) there's also just a lot of file additions so it's not that difficult.

Is that ok?

**Comment by tanghus** [comment]
http://www.git-scm.com/book/en/Git-Tools-Rewriting-History#Squashing-Commits

**Comment by babelouest** [comment]
@tanghus , try with the last commit for your bugs: https://github.com/owncloud/contacts/commit/3b516a1808aa95f378c2b467ea2e724b2860d849
Should be fixed.

@jancborchardt , using your technique or tabghus' technique, either way I suggest I'll fix that when the new import interface will pass your tests, so it will be done once ;)
Do yo agree ?

**Comment by tanghus** [comment]
You mean dfba196 right? ;) I'll give it a go tomorrow.

**Comment by babelouest** [comment]
@tanghus , yup ;)

**Comment by tanghus** [comment]
@babelouest @jancborchardt I made so many changes when cleaning up the code in master that this wasn't mergable anymore. A cleaned up version is now in #443

**Pull request closed without merging**